File: Python/0211-add-and-search-word-data-structure-design.py
# Time:  O(min(n, h)), per operation
# Space: O(min(n, h))

class TrieNode:
    def __int__(self):
        self.isCompleteWord = False
        self.children = {}

class WordDictionary:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.root = TrieNode()

    def addWord(self, word: str) -> None:
        """
        Adds a word into the data structure.
        """
        cur = self.root
        for c in word:
            if c not in cur.children:
                cur.children[c] = TrieNode()
            cur = cur.children[c]
        cur.isCompleteWord = True

    def search(self, word: str) -> bool:
        """
        Returns if the word is in the data structure. A word could contain the dot character '.' to represent any one letter.
        """
        return self.searchHelper(word, 0, self.root)

    def searchHelper(self, word, start, cur):
        if start == len(word):
            return cur.isCompleteWord
        if word[start] in cur.children:
            return self.searchHelper(word, start+1, cur.children[word[start]])
        elif word[start] == '.':
            for c in cur.children:
                if self.searchHelper(word, start+1, cur.children[c]):
                    return True

        return False


# Your WordDictionary object will be instantiated and called as such:
# obj = WordDictionary()
# obj.addWord(word)
# param_2 = obj.search(word)

# Notes: AttributeError: TrieNode object has no attribute children (Line 23)

File: C++/0211-add-and-search-word-data-structure-design.cpp
// Time: O(min(n, h)) for search() if no '.' else O(26^n), O(n) for addWord()
// Space: O(min(n, h)) for search(), O(1) for addWord()

struct TrieNode {
        bool isCompleteWord = false;
        unordered_map<char, TrieNode *> children;
};

class WordDictionary {
private:
    TrieNode *root;
public:
    /** Initialize your data structure here. */
    WordDictionary() {
        root = new TrieNode();
        root->isCompleteWord = true;
    }

    /** Adds a word into the data structure. */
    void addWord(string word) {
        auto *cur = root;
        for (const auto &c: word) {
            // no match
            if (cur->children.find(c) == cur->children.cend())
                cur->children[c] = new TrieNode;
            cur = cur->children[c];
        }
        cur->isCompleteWord = true;
    }

    /** Returns if the word is in the data structure. A word could contain the
    dot character '.' to represent any one letter. */
    bool search(string word) {
        return searchHelper(word, 0, root);
    }

    bool searchHelper(string word, int start, TrieNode *cur) {
        if (start == word.length())
            return cur->isCompleteWord;

        if (cur->children.find(word[start]) != cur->children.end()) {
            return searchHelper(word, start + 1, cur->children[word[start]]);
        } else if (word[start] == '.') {
            for (const auto &i: cur->children) {
                if (searchHelper(word, start + 1, i.second))
                    return true;
            }
        }
        return false;
    }
};

/**
 * Your WordDictionary object will be instantiated and called as such:
 * WordDictionary* obj = new WordDictionary();
 * obj->addWord(word);
 * bool param_2 = obj->search(word);
 */

 /* Notes: Space required for search() is O(min(n, h)) because if n (string len)
 is less than h then our stack only goes upto height n */
